tests/test_math.h

```diff
@@ -31,11 +31,55 @@
 #ifndef TEST_MATH_H
 #define TEST_MATH_H
 
-#include "core/os/main_loop.h"
+#include "core/math/math_defs.h"
+#include "core/math/math_funcs.h"
+#include "core/math/transform.h"
+#include "core/os/os.h"
+
+#include <math.h>
+#include <stdio.h>
+
+#include "tests/test_macros.h"
 
 namespace TestMath {
 
 MainLoop *test();
 }
 
+TEST_CASE("[Transform] Rotate around global origin") {
+	// Start with the default orientation, but not centered on the origin.
+	// Rotating should rotate both our basis and the origin.
+	Transform transform = Transform();
+	transform.origin = Vector3(0, 0, 1);
+
+	Transform expected = Transform();
+	expected.origin = Vector3(0, 0, -1);
+	expected.basis.set_axis(0, Vector3(-1, 0, 0));
+	expected.basis.set_axis(2, Vector3(0, 0, -1));
+
+	Transform rotated_transform = transform.rotated(Vector3(0, 1, 0), Math_PI);
+	REQUIRE(rotated_transform.is_equal_approx(expected));
+
+	// Make sure that the rotated transform isn't sharing references with the original transform.
+	REQUIRE(&rotated_transform.basis != &transform.basis);
+	REQUIRE(&rotated_transform.origin != &transform.origin);
```

Comment:
While the first `REQUIRE` is reasonable, the successive assertions depend on the previous one. Better do this:

```suggestion
	Transform rotated_transform = transform.rotated(Vector3(0, 1, 0), Math_PI);
	REQUIRE(rotated_transform.is_equal_approx(expected));

	// Make sure that the rotated transform isn't sharing references with the original transform.
	CHECK(&rotated_transform.basis != &transform.basis);
	CHECK(&rotated_transform.origin != &transform.origin);
```

This way, both "basis" and "origin" will be tested (without immediately exiting the test case if the first one fails).

Same for the test case above.

Reply:
Seems reasonable, I'll change the latter REQUIREs to CHECKs. 

As far as referencing with `&`, this may just be paranoia due to my lack of C++ experience. But technically the `==` operator can be overloaded, right?

Reply:
Looking more closely now, in this case `&` will take a pointer, so this will always "falsely" succeed because this compares pointers rather than values, because as far as I know `rotated_transform` is a modified copy of `transform`, so `&` is not suitable in this case.

So this should rather be:
```suggestion
	Transform rotated_transform = transform.rotated(Vector3(0, 1, 0), Math_PI);
	REQUIRE(rotated_transform.is_equal_approx(expected));

	// Make sure that the rotated transform isn't sharing references with the original transform.
	CHECK(rotated_transform.basis != transform.basis);
	CHECK(rotated_transform.origin != transform.origin);
```

> But technically the `==` operator can be overloaded, right?

Yes, this is done for a lot of existing core types in Godot, so usually you don't have to worry about this and just to equality checks. If those fail due to floating point accumulation, `is_equal_approx` is also implemented for built-in types.

Reply:
I think the pointer memory addresses are what I wanted to check; I thought I could make sure that `transform.rotated()` returns an object that doesn't share important references with the old object. I.e. that the following is the case -

`Transform transform = Transform();`
`Transform rotated = transform.rotated(...)`;
`transform.origin.x += 1;`
`//rotated's origin should still be on the origin, not on (1,0,0)`

But that's maybe not necessary. In general, as a new Godot user, I'm not always clear on whether a function will modify the existing object or return a new object.